man/man1/major_minor.1
.TH MAJOR_MINOR 1 
.ad
.fi
.SH NAME
major_minor
\-
PERL start engine
.SH SYNOPSIS
.na
.nf
\fBeval\fR `\fBmajor_minor\fR`;

\fB$major_number = &dev_major($devnum);\fR

\fB$minor_number = &dev_minor($devnum);\fR
.SH DESCRIPTION
.ad
.fi
\fBmajor_minor\fR emits two PERL routines, \fBdev_major\fR()
and \fBdev_minor\fR() that take a device number as returned
by \fBstat\fR() and that break the number up into the device
major and minor number, respectively.

For example, the FreeBSD 2.2 output is:

.DS
sub dev_major { local($dev) = @_; ($dev >> 8) & 0xff; };

sub dev_minor { local($dev) = @_; ($dev & 0xffff00ff); };

1;
.DE

In principle, this information could be obtained by parsing
system include files, but a little C program is more robust.
.SH HISTORY
.na
.nf
.fi
.ad
A similar trick was used first with the SATAN program.
.SH LICENSE
.na
.nf
This software is distributed under the IBM Public License.
.SH AUTHOR(S)
.na
.nf
Wietse Venema
IBM T.J. Watson Research
P.O. Box 704
Yorktown Heights, NY 10598, USA


CHANGES
Sat Aug 29 2009

o	grave-robber did not record the initial directory in the body file.

o	Bit-rot: FreeBSD 7 support, including ``struct acct'' changes.

o	Bit-rot: lastcomm format strings.

Sun Jan 11 20:01:27 EST 2009

o	Bit-rot: OpenBSD 4.x support (still TODO: file system tools).

Thu Jun  7 10:54:44 EDT 2007

o	Bit-rot: FreeBSD 6.x support.

Mon Apr  9 09:36:31 EDT 2007

o	Bit-rot: the 64-bit workaround for LINUX lseek() is no
	longer needed, and it didn't compile on some systems.

Thu Jan 11 16:10:29 EST 2007

o	Bugfix: the timeout command always reported zero exit status
	after voluntary child process exit.

Mon Jun 26 18:44:17 EDT 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Fri Mar 17 09:25:57 EST 2006

o	Bit-rot: update #include statements because code either
	no longer compiled or compiled with warnings.

Tue Jan  6 12:25:32 EST 2004

o       Workaround: mactime treated "," as a special character in
	file names, now it's 0xff.

o	Support for ext[23]fs inode file sizes > 32bits.

o	Bugfix: ils file sizes larger than unsigned long.

Tue Oct 14 18:23:49 EDT 2003

o       pcat now fully supports FreeBSD systems that do not have
	/proc mounted (it's no longer mounted by default with
	FreeBSD 5). pcat tries to use /proc if it can, and uses
	ptrace() and gropes kernel memory if it has to.  Operation
	without /proc requires super-user privileges.

Thu Oct  2 09:14:00 EDT 2003

o       Completed support for UFS1 and UFS2 in FreeBSD 5.x. However,
	the pcat command still requires that the /proc file system
	is mounted. This will be fixed later.

Sat Aug 30 19:21:29 EDT 2003

o	Preliminary FreeBSD 5.0 port. UFS2 support is still to be done.

Fri Aug  1 10:07:39 EDT 2003

o	RedHat 9.0 fix by Florin Andrei for the file command.

Sat Oct  5 13:47:29 EDT 2002

o	Restored tctutils compatibility, which broke with 20020916.

Mon Sep 16 14:39:09 EDT 2002

o       The lowest inode examined is now the lowest possible inode
	number instead of ROOTINUM.

Fri Mar 22 14:41:32 EST 2002

o	Added support for OpenBSD 3.0.

Wed Dec 12 13:21:00 EST 2001

o	tm_print() and tm_string() did unnecessary %HEX expansion
	of $ and ^ characters (Philippe Bourgeois).

o       mactime mis-sorted dates before/after early September 2001
	(Philippe Bourgeois, Brian Carrier).

Thu Oct  4 15:48:42 EDT 2001

o       Linux can handle files > 2Gbyte, but only if we specify
	extra compiler flags.

o	Solaris 8/64bit needs an extra include file in pcat.c.

Sun Sep  9 17:23:13 EDT 2001

o       Discovered to my horror that unrm on ext2fs can have an
	off-by-one error when testing a bit flag in the data block
	allocation bitmaps - and this problem is dependent on the
	Linux version being used.

o       Made the timeout command a little more verbose so it is
	less of a mystery when the child process is killed.

Thu Sep  6 13:51:00 EDT 2001

o	New grave-robber -I option for better control of icat/pcat.

Mon May 21 10:11:19 EDT 2001

o       findkey command-line error check was too conservative,
	and disallowed threshold of 256 different byte values
	(Ovanes Manucharyan).

Mon May 14 13:11:54 EDT 2001

o	Patch 06 broke lib/pass.cache.pl, which broke mactime.

o	Patch 06 did not update the patchlevel file.

Fri Mar 16 14:25:11 EST 2001

o	Fixed rpm root directory when running on a corpse.
	Marco Walter (SUN).

o	Fixed file(1) output parsing problem with multiple :
	characters.  Marco Walter (SUN)

Thu Feb 22 11:16:40 EST 2001

o	Patch by Brian Carrier (Purdue) so that fs_read_random()
	properly maintains file offset information.

Thu Feb  8 10:40:02 EST 2001

o	grave-robber dies silently when no group file is found.
	This bites when examining a partial machine corpse
	(bin/grave-robber). Klaus-Peter Kossakowski.

o	The "do first looks" code did not relocate to $TCT_HOME
	(lib/tree.pl). Klaus-Peter Kossakowski.

Tue Jan 23 18:44:36 EST 2001

o       The grave-robber did not correctly handle multi-line df
	output records (ps_spy.pl, suck_free_inodes.pl).

o       The grave-robber mounted corpse inode scan looked only at
	the corpse's "root" file system (suck_free_inodes.pl).

Mon Jan 22 11:23:32 EST 2001

o       grave-robber enabled the default -lPO flags when -c was
	specified, causing the program to gripe and exit.
	
Tue Sep 26 10:55:24 EDT 2000

o       Fixed a bug in fs_read_block() that caused icat etc. to
	remember the wrong file seek position, so that icat etc.
	would call lseek() too often. The other possibility, not
	calling lseek() when it should be called, did not happen
	due to an implementation artefact of the TCT utilities.
	Reported by Brian Carrier, Purdue University.

	Updated help-recovering-file document by Dan.

Sat Aug 12 10:51:16 EDT 2000

o       Added -V option to direct icat/ils/unrm for verbose output
	to stdout.  The -v option now sends verbose logging to stderr.

Tue Aug  8 11:46:19 PDT 2000

o	created "help-recovering-file", added note to README.FIRST

Sat Aug  5 11:34:26 PDT 2000

o	mactime time range repaired, fixed docs, -l option slightly
	fixed... perhaps deprecation in the future?

Fri Aug  4 20:06:15 EDT 2000

o	ils and ils2mac now include device names for more
	readable mactime reports of dead inodes.

o	mactime could not run with current working directory
	different from $TCT_HOME.

o	mactime -d could destroy an existing $DATA/body file.

o	mactime -d turned on verbosity, instead of letting -v do that.

o	mactime -s (suid) flag used an obsolete method of getting data.

o	ils2mac broke because perl split() can't see the difference
	between empty last fields and missing last fields.

o	mactime mistakenly skipped files with all-zero time stamps.

o	ils2mac now emits ls -l like info for mactime compatibility.

Tue Aug  1 13:50:20 PDT 2000

o	Prepend our directory list to @INC rather than clobbering it.

Mon Jul 31 17:37:35 PDT 2000

o	removed mention of sans.org from help-when-broken-into


src/major_minor/major_minor.c
/*++
/* NAME
/*	major_minor 1
/* SUMMARY
/*	PERL start engine
/* SYNOPSIS
/*	\fBeval\fR `\fBmajor_minor\fR`;
/*
/*	\fB$major_number = &dev_major($devnum);\fR
/*
/*	\fB$minor_number = &dev_minor($devnum);\fR
/* DESCRIPTION
/*	\fBmajor_minor\fR emits two PERL routines, \fBdev_major\fR()
/*	and \fBdev_minor\fR() that take a device number as returned
/*	by \fBstat\fR() and that break the number up into the device
/*	major and minor number, respectively.
/*
/*	For example, the FreeBSD 2.2 output is:
/*
/* .DS
/*	sub dev_major { local($dev) = @_; ($dev >> 8) & 0xff; };
/*
/*	sub dev_minor { local($dev) = @_; ($dev & 0xffff00ff); };
/*
/*	1;
/* .DE
/*
/*	In principle, this information could be obtained by parsing
/*	system include files, but a little C program is more robust.
/* HISTORY
/* .fi
/* .ad
/*	A similar trick was used first with the SATAN program.
/* LICENSE
/*	This software is distributed under the IBM Public License.
/* AUTHOR(S)
/*	Wietse Venema
/*	IBM T.J. Watson Research
/*	P.O. Box 704
/*	Yorktown Heights, NY 10598, USA
/*--*/

#include <sys/types.h>
#ifdef SUNOS5
#include <sys/mkdev.h>
#define TEMPLATE (NODEV>>1)
#endif
#ifdef LINUX2
#include <sys/sysmacros.h>
#endif
#ifndef TEMPLATE
#define TEMPLATE 0xffffffff
#endif

#include <stdio.h>

int     main(int argc, char **argv)
{
    unsigned foo = TEMPLATE;
    unsigned min = minor(foo);
    unsigned shift;
    unsigned mask;

    for (mask = ~min, shift = 0; (mask & 1) == 0; shift++, mask >>= 1)
	 /* void */ ;

    printf("sub dev_major { local($dev) = @_; ($dev >> %d) & 0x%x; };\n\n",
	   shift, major(foo));
    printf("sub dev_minor { local($dev) = @_; ($dev & 0x%x); };\n\n",
	   min);
    printf("1;\n");
    return (0);
}
